src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php

```diff
@@ -230,114 +190,106 @@ public function putAction($id, Request $request)
+     * for children of a category.
+     * If expandIds is set, the paths to the categories which are assigned to the ids are expanded. Categories are
+     * only expanded if they are descendants of the root level and the do not affect the limit or pagination.
      *
-     * @throws MissingParameterException iff the locale is missing in the request
+     * @param Request $request
+     * @param $locale
+     * @param null $parentId
+     * @param bool $disableLimit
+     * @param array $expandIds
      *
      * @return CategoryListRepresentation
      */
-    protected function getCategoryListRepresentation(Request $request, $parentKey = null)
+    protected function getListRepresentation(Request $request, $locale, $parentId = null, $disableLimit = false, $expandIds = [])
     {
-        /** @var RestHelperInterface $restHelper */
-        $restHelper = $this->getRestHelper();
-
-        /** @var DoctrineListBuilderFactory $factory */
-        $factory = $this->get('sulu_core.doctrine_list_builder_factory');
+        $listBuilder = $this->initializeListBuilder($locale);
 
-        $listBuilder = $factory->create(self::$entityName);
-        $locale = $this->getRequestParameter($request, 'locale', true);
-        $fieldDescriptors = $this->getManager()->getFieldDescriptors($locale);
-        $listBuilder->sort($fieldDescriptors['depth']);
+        // if search is set, return all matching categories. else return only level below parent in category tree
+        if (!$request->get('search')) {
+            $listBuilder->where($this->getCategoryManager()->getFieldDescriptor($locale, 'parent'), $parentId);
+        }
 
-        $restHelper->initializeListBuilder($listBuilder, $fieldDescriptors);
+        // need to disable default limit because some frontend components are not paginated
+        if ($disableLimit || !$request->get('limit')) {
+            $listBuilder->limit(null);
+        }
 
-        $listBuilder->addSelectField($fieldDescriptors['depth']);
-        $listBuilder->addSelectField($fieldDescriptors['parent']);
-        $listBuilder->addSelectField($fieldDescriptors['locale']);
-        $listBuilder->addSelectField($fieldDescriptors['defaultLocale']);
-        $listBuilder->addSelectField($fieldDescriptors['lft']);
-        $listBuilder->addSelectField($fieldDescriptors['rgt']);
+        $results = $listBuilder->execute();
 
-        if ($parentKey !== null) {
-            $this->addParentSelector($parentKey, $listBuilder);
+        if ($expandIds) {
+            // expand the paths to the categories which are assigned to the given expandIds.
+            // we need to do this in a second query, because we need to expand children of the current page only.
+            $resultIds = array_map(
+                function ($category) {
+                    return $category['id'];
+                },
+                $results
```

Comment:
Well, pagination isn't the most complicated thing we've ever built... And apart from that I am not sure if it really makes sense anyway, maybe we can remove it? /cc @chirimoya 

Apart from that we have a complexity of O(n), since the time is not fixed. I would still go with the separate query, and not using the listbuilder, which was not build for only retrieving the IDs of something and does massive overfetching in this case.